### PR TITLE
Add population slider with tests

### PR DESCRIPTION
**Pull request by tunglinn**
Two sliders stacked on top of each other.
In order for it to look visually smooth, the length of two sliders are adjusted as the slider heads move.

image

### CODE REVIEW

File: src/css/_slider.scss [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  position: absolute;
+  z-index: 1100;
+  display: flex;
+  flex-direction: column;
+  align-items: center;
+  top: 10em;
+  left: 1em;
+}
+.min-max-slider {
+  width: 360px;
```
Comment (tunglinn) [inline src/css/_slider.scss:16]: Will change once in a popup

File: src/js/slider.js [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+  // Writing and inserting header label
+  var lower = document.querySelector(".lower");
+  var upper = document.querySelector(".upper");
+  //   lower.classList.add("lower", "value");
+  //   upper.classList.add("upper", "value");
+  lower.appendChild(document.createTextNode(rangemin));
+  upper.appendChild(document.createTextNode(rangemax));
+  //   slider.insertBefore(lower, leftSlider.previousElementSibling);
+  //   slider.insertBefore(upper, leftSlider.previousElementSibling);
```
Comment (tunglinn) [inline src/js/slider.js:93]: will remove along with next edits

File: src/css/_slider.scss [hunk 3]
```diff
@@ -0,0 +1,77 @@
+// uncomment to visualize slider mechanism
+// input {
+//   border: solid;
+// }
```
Comment (Eric-Arellano) [inline src/css/_slider.scss:4]: This is excellent. Very useful for future developers :)

File: src/css/_slider.scss [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+  z-index: 1100;
+  display: flex;
+  flex-direction: column;
+  align-items: center;
+  top: 10em;
+  left: 1em;
+}
+.min-max-slider {
+  width: 360px;
+  //   text-align: center;
```
Comment (Eric-Arellano) [inline src/css/_slider.scss:17]: Dead code should be removed
```suggestion
```

File: src/css/_slider.scss [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+  height: 2em;
+  display: flex;
+  justify-content: space-between;
+  // margin-bottom: 50px;
+  // top: 10em;
+  // left: 1rem;
```
Comment (Eric-Arellano) [inline src/css/_slider.scss:23]:
```suggestion
```

File: src/css/_slider.scss [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+}
+.min-max-slider > label {
+  display: none;
+}
```
Comment (Eric-Arellano) [inline src/css/_slider.scss:27]: This is better as a nested rule in the `.min-max-slider`, with `> label` as the selector. 

But also, is it necessary to set this? Why hide the label?

File: tests/app/filter.test.js [hunk 7]
```diff
@@ -31,3 +31,20 @@ test("scope filter updates markers", async ({ page }) => {
...
+  const right = await page.$(".right-slider");
+
+  const testSlider = async (slider, moveTo) => {
+    const before = await getNumCities(page);
+    await slider.fill(moveTo);
+    const after = await getNumCities(page);
+    expect(before > after).toBe(true);
+  };
+
+  await testSlider(left, "5");
```
Comment (Eric-Arellano) [inline tests/app/filter.test.js:47]: What does this number 5 correspond to? Like what is the unit? You could address this by giving a better name to `moveTo` so it's "self-documenting" code, like `moveToValue` or `movePixelsAmount`.

File: tests/app/filter.test.js [hunk 8]
```diff
@@ -31,3 +31,20 @@ test("scope filter updates markers", async ({ page }) => {
...
+test("population filter updates markers", async ({ page }) => {
+  await loadMap(page);
+  const left = await page.$(".left-slider");
+  const right = await page.$(".right-slider");
+
+  const testSlider = async (slider, moveTo) => {
+    const before = await getNumCities(page);
+    await slider.fill(moveTo);
+    const after = await getNumCities(page);
+    expect(before > after).toBe(true);
```
Comment (Eric-Arellano) [inline tests/app/filter.test.js:44]: I think it's probably better to have a range of what we expect the values to be, like you did with the filter. That helps us to catch more edge cases with issues in the implementation.

Then, with that change, you can test some more edge cases like:

* Only 100 is selected
* 100, 500, and 1k are selected
* Everything but 100, 500, and 1k are selected

Having more edge cases covered does make the test slower to run. But that's faster than us accidentally breaking the functionality and not realizing!

File: src/js/slider.js [hunk 9]
```diff
@@ -0,0 +1,150 @@
+import { changeSelectedMarkers } from "./filter";
```
Comment (Eric-Arellano) [inline src/js/slider.js:1]: maybe call the file `populationSlider.js` so that it's more self-documenting what this thing is about. Ditto the other files.

File: src/css/_slider.scss [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+span.value {
```
Comment (Eric-Arellano) [inline src/css/_slider.scss:28]: This is a bad selector because it's too general. It impacts every single `span` ever. You want to nest it in `.slider-legend`.

File: index.html [hunk 11]
```diff
@@ -117,6 +117,34 @@ <h3>Concept and Initial Report</h3>
       <option value="Transit Oriented">Transit Oriented</option>
       <option value="Main Street/Special">Main Street/Special</option>
     </select>
+    <div class="population-slider">
```
Comment (Eric-Arellano) [inline index.html:120]: I recommend calling this `population-slider-container` to make clear it is not the slider itself. That helps to clarify the difference with `min-max-slider`, which you could rename to `population-slider`. Then call `slider-label` `population-slider-label` to be consistent.

File: src/css/_slider.scss [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+  font-weight: bold;
+  display: inline-block;
+}
+.slider-legend {
+  display: flex;
+  justify-content: space-between;
+  width: 100%;
+  padding: 0;
+}
+.slider-legend span {
```
Comment (Eric-Arellano) [inline src/css/_slider.scss:39]: This should be nested

File: src/css/_slider.scss [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+  opacity: 1;
+  flex: 1;
+  text-align: center;
+}
+.min-max-slider > input {
+  cursor: pointer;
+  position: absolute;
+}
+
+/* webkit specific styling */
```
Comment (Eric-Arellano) [inline src/css/_slider.scss:50]: This comment is being applied to all of the rule set, but it only actually matters to line 52, so it should be moved closer. But, in reality, you can remove it -- the name `-webkit-appearance` already makes it obvious.

Btw, with Sass, you can use `//` for comments.

File: src/css/_slider.scss [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+.min-max-slider > input {
+  -webkit-appearance: none;
+  outline: none !important;
```
Comment (Eric-Arellano) [inline src/css/_slider.scss:53]: Why the `!important`? Those are best to ignore when possible.

File: src/css/_slider.scss [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+    transparent 60%,
+    transparent 100%
+  );
+}
+.min-max-slider > input::-webkit-slider-thumb {
+  -webkit-appearance: none; /* Override default look */
+  appearance: none;
+  width: 14px; /* Set a specific slider handle width */
+  height: 14px; /* Slider handle height */
+  background: #eee; /* Green background */
```
Comment (Eric-Arellano) [inline src/css/_slider.scss:70]: Better is to use `green`, if that's the same as `#eee`.

File: src/css/_slider.scss [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+    transparent 0%,
+    transparent 30%,
+    silver 30%,
+    silver 60%,
+    transparent 60%,
+    transparent 100%
+  );
+}
+.min-max-slider > input::-webkit-slider-thumb {
+  -webkit-appearance: none; /* Override default look */
```
Comment (Eric-Arellano) [inline src/css/_slider.scss:66]: You can use `//` for all these comments

File: src/js/slider.js [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+];
+
+const numInterval = [
+  100, 500, 1000, 5000, 10000, 50000, 100000, 500000, 1000000, 5000000,
+  10000000, 50000000,
+];
+
+const draw = (slider, low, high) => {
+  // Setting vars from top to bottom.
+  var lower = document.querySelector(".lower");
```
Comment (Eric-Arellano) [inline src/js/slider.js:26]: We want to use `const` for each of these. `var` is the old API and has a lot of weird edge cases.

File: src/js/slider.js [hunk 18]
```diff
@@ -0,0 +1,150 @@
...
+  var upper = document.querySelector(".upper");
+  var leftSlider = slider.querySelector(".left-slider"); // left slider
+  var rightSlider = slider.querySelector(".right-slider"); // right slider
```
Comment (Eric-Arellano) [inline src/js/slider.js:29]: These comments aren't useful. Ditto a few others in the file. The code is already self documenting.

File: index.html [hunk 19]
```diff
@@ -117,6 +117,34 @@ <h3>Concept and Initial Report</h3>
...
+      </div>
+      <div class="min-max-slider" data-legendnum="12">
+        <input
+          id="left-slider"
+          class="left-slider"
+          name="min"
+          type="range"
+          step="0.5"
+          min="0"
+          max="11"
```
Comment (Eric-Arellano) [inline index.html:134]: How do these get determined? If we had fewer options per https://github.com/ParkingReformNetwork/reform-map/issues/231, would these change?

**Review by Eric-Arellano — COMMENTED** [review]
Nice, this works great!

File: tests/app/filter.test.js (hunk 8 above)
Comment (tunglinn) [reply tests/app/filter.test.js]: Still thinking about how to implement this. Like a data structure that will easily store and retrieve the differences

File: src/css/_population-slider.scss [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+  height: 1.7em;
+  font-weight: bold;
+  display: inline-block;
+}
+.slider-legend {
+  display: flex;
+  justify-content: space-between;
+  width: 100%;
+  padding: 0;
+  span {
```
Comment (Eric-Arellano) [inline src/css/_population-slider.scss:31]:
```suggestion
  padding: 0;

  span {
```

File: src/css/_population-slider.scss [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+    flex: 1;
+    text-align: center;
+  }
+}
+.population-slider > input {
+  cursor: pointer;
+  position: absolute;
+}
+
+.population-slider > input {
```
Comment (Eric-Arellano) [inline src/css/_population-slider.scss:43]: Same selector is repeated twice
```suggestion
```

File: src/css/_population-slider.scss [hunk 22]
```diff
@@ -0,0 +1,69 @@
...
+    transparent 60%,
+    transparent 100%
+  );
+}
+.population-slider > input::-webkit-slider-thumb {
+  -webkit-appearance: none; // Override default look
+  appearance: none;
+  width: 14px; // Set a specific slider handle width
+  height: 14px; // Slider handle height
+  background: white; // white, solid background
```
Comment (Eric-Arellano) [inline src/css/_population-slider.scss:62]:
```suggestion
  width: 14px;
  height: 14px;
  background: white;
```

File: src/css/_population-slider.scss [hunk 23]
```diff
@@ -0,0 +1,69 @@
...
+  cursor: pointer;
+  border: 1px solid gray;
+  border-radius: 100%;
+}
+.population-slider > input::-webkit-slider-runnable-track {
```
Comment (Eric-Arellano) [inline src/css/_population-slider.scss:67]: It's good to have whitespace between each selector, like above this line

File: src/js/populationSlider.ts [hunk 24]
```diff
@@ -0,0 +1,162 @@
...
+  // Set attributes before drawing
+  leftSlider.setAttribute("data-value", leftValue);
+  rightSlider.setAttribute("data-value", rightValue);
+  leftSlider.value = leftValue;
+  rightSlider.value = rightValue;
+
+  const min = numInterval[leftValue];
+  const max = numInterval[rightValue];
+
+  const inBetween = (x, low, high) => {
```
Comment (Eric-Arellano) [inline src/js/populationSlider.ts:144]: This would benefit from type hints

File: src/js/populationSlider.ts [hunk 25]
```diff
@@ -0,0 +1,162 @@
...
+  };
+
+  changeSelectedMarkers(markerGroup, citiesToMarkers, (cityState) =>
+    inBetween(parseInt(data[cityState]["population"]), min, max)
+  );
+
+  draw(slider, stringIntervals[leftValue], stringIntervals[rightValue]);
+};
+
+const setUpSlider = (markerGroup, citiesToMarkers, data) => {
```
Comment (Eric-Arellano) [inline src/js/populationSlider.ts:155]: Ditto adding type hints

File: src/js/populationSlider.ts [hunk 26]
```diff
@@ -0,0 +1,162 @@
...
+    updateExponential(rightSlider, markerGroup, citiesToMarkers, data);
+  });
+};
+
+const updateExponential = (
+  el: HTMLInputElement,
+  markerGroup: FeatureGroup,
+  citiesToMarkers: Record<CityId, CircleMarker>,
+  data: Record<CityId, CityEntry>
+) => {
```
Comment (Eric-Arellano) [inline src/js/populationSlider.ts:127]: Return type is `void`

**Review by Eric-Arellano — APPROVED** [review]
